### ISSUE TITLE
Discussion: Deprecate the term "dummy application"

### ISSUE THREAD

**Issue by elwayman02** [emberjs/rfcs#718]
Note: I made this issue as a pre-RFC, because I'd like to spark some discussion of the topic before moving into an actual RFC, as I don't consider the idea fully-formed as of yet, particularly in terms of the impact and where things need to change.

# Background

Historically, Ember addons have referred to an application living within the project as the ["dummy application"](https://cli.emberjs.com/release/writing-addons/#testsdummy). As noted in the official ember docs, this app exists for the purpose of testing the addon itself. The application exists at the sub-directory `tests/dummy/` within the addon's structure. The application has also traditionally been titled "Dummy" by default, in the HTML head.

# Problem Statement

The term "dummy" is not ideal for a number of reasons:

## Descriptiveness

"Dummy" does not accurately describe the purpose of the application that lives within addons. As the community has evolved and this tooling has become more sophisticated, there are three primary use-cases that have evolved for most projects:

1. Testing - The application is used to test functionality from the addon as a supplement to automated testing
2. Documentation - The application is used for generating a deployable documentation site that consumers of the addon can use as a resource. Some examples of projects focused on this use-case are [field-guide](https://github.com/empress/field-guide/) and [ember-cli-addon-docs](https://github.com/ember-learn/ember-cli-addon-docs).
3. Demonstration - The application is used to show live demos of the addon's functionality, often in conjunction with Testing and Documentation purposes.

## Ableist Language

The word "dummy" is [rooted](https://www.etymonline.com/word/dummy) in [ableism](https://www.etymonline.com/word/dumb), and as such is a poor example of the inclusive language we strive for within the Ember community. While our use of it is not unique within the tech industry, we would also not be alone in working toward removing such language from our vernacular as developers. For example, Google's [developer documentation style guide](https://developers.google.com/style/inclusive-documentation) specifically calls out the phrase "dummy variable" as an example of ableist language that should replaced with inclusive alternatives like "placeholder variable".

# Solution

I would propose that we find more inclusive language to replace "dummy", both in our code and the way we describe this addon feature. My personal preference is "Demo Application", as I find that to be generally inclusive of all three purposes described above, since Demos often exist in both test applications and documentation. I'm open to other suggestions, however!

## Documentation

We would need to update the documentation for ember-cli and any guides that reference the "dummy app" to replace it with the new terminology. One such place where said documentation exists is below:

https://cli.emberjs.com/release/writing-addons

## Code

The `tests/dummy/` sub-directory would be renamed to reflect the more inclusive term, and any tooling that hooks into this application needs to be updated to the new file locations. These changes would take place in the addon blueprint as well as ember-cli itself.

## Migration Path

We'd need to prepare tooling to make it simple for existing addons to migrate to the new file structure. Some projects may have custom behavior that interacts with this application, so we'd need to document these changes and try to capture some common use-cases to make it easy to update.

It would also be wise to maintain backwards-compatible functionality for ember-cli to continue working with addons that haven't updated to the new file structure, until a point in the future where that deprecated functionality can be safely removed without breaking a large number of community projects.

# TODO

- [ ] Search major ember projects (including community ones) for references to "dummy"
- [ ] Identify non-breaking migration path

**Comment by knownasilya** [comment]
👍🏼 on the rename. I wonder if it makes more sense to convert addon folder structure to a monorepo with the demo a sibling of the addon. This way others can be created for testing, and docs. Basically deprecating the "dummy" app as-is currently and just generating a full app.

**Comment by elwayman02** [comment]
Honestly, that's also a great idea @knownasilya, though I'm not sure if we'd want that part of this (eventual) RFC or as a separate RFC.

**Comment by jenweber** [comment]
I like this general idea. We have many other good words available to us. Some that come to mind in addition to those mentioned already are:

- sample-app
- testing-app
- example-app

An RFC for this should account for the glide path for tooling that relies on the location of this folder. Ideally it would be aliased somehow so that it does not need to be a breaking change.

**Comment by rwjblue** [comment]
I agree that "dummy" isn't the best name and we can find another one that works better (roughly what @jenweber said), but it definitely isn't an ableist term in this context. It is referring to the idea of (from the [etymology link](https://www.etymonline.com/word/dummy) above):

> counterfeit object, something that imitates a reality for mechanical purposes

Something akin to a [crash test dummies](https://en.wikipedia.org/wiki/Crash_test_dummy).

**Comment by rwjblue** [comment]
I don't particularly think that a name change for the sake of avoiding specific words (that actually are the **correct** words in this context) is worth the churn on our users.

If we can find a way to accommodate a better name as part of a broader goal, that seems totally fine. For example, "soon"(ish) we will have the opportunity to rethink the addon authoring format (as part of the Embroider v2 addon work), at that point I think it would be a great idea to incorporate something like @knownasilya's idea (allowing for multiple test applications) and a name change is perfectly logical in that context.

**Comment by hjdivad** [comment]
The monorepo approach helps with a particular pain point of testing addons that behave differently depending on peer dependencies.  It also helps make the example application a lot easier to understand as there's much less implicit translation a user has to do.

**Comment by elwayman02** [comment]
> I agree that "dummy" isn't the best name and we can find another one that works better (roughly what @jenweber said), but it definitely isn't an ableist term in this context. It is referring to the idea of (from the [etymology link](https://www.etymonline.com/word/dummy) above):
> 
> > counterfeit object, something that imitates a reality for mechanical purposes
> 
> Something akin to a [crash test dummies](https://en.wikipedia.org/wiki/Crash_test_dummy).

I don't actually agree with this. While the intention might have been to draw inspiration from something like "crash test dummy", we have to realize that even that term was drawn from somewhere else originally. "Dummy" originally meant "someone who can't speak", and then its meaning morphed further into definitions like "a substitute for the real thing", which is what eventually led to the "counterfeit object" definition. The fact that the language evolved over time to be more expansive in its application doesn't change that it's rooted in ableism.

Furthermore, the implied context that you and I understand doesn't necessarily mean that it is well understood by all within the community or those who may encounter our community from the outside. The fact that the term "dummy application" does not have an immediately obvious context and isn't very descriptive makes this worse, because it allows a negative perception to exist from the use of the word.

Ultimately, though, I would refer back to Google's style guide. I don't think our use of the word "dummy" varies significantly from the example they used, not in any way that is meaningfully important. I don't find a lot of value in arguing "well, the way we *intend* this word isn't ableist, so it's fine to leave it that way". Which isn't exactly what you said, but it's easy for other people to come away *thinking* that after reading your comment.

**Comment by mydea** [comment]
For what it's worth, the term "dummy app" never was intuitive or made much sense to me as a non-native English speaker. I also always found it difficult to explain it and it wasn't obvious/intuitive to anyone I taught this (all non-English native speakers, so might be a language thing). 

I would find e.g. `demo` or `example` much more intuitive and self-explanatory. For me, changing the name `dummy` to something different would be a pretty easy and helpful change on it's own, and make learning how these things work easier for newcomers as well.

Regarding the monorepo approach, one thing this might not cover very well is using stuff from the dummy app in acceptance tests. E.g. I have often written acceptance tests in addons testing something layed out in the dummy app, especially when it includes testing routing behavior and similar things. Sure, in such case I guess you could also write this as tests directly in the demo application I guess, but this would also be rather cumbersome. Just as a point to consider.

**Comment by nightire** [comment]
> Sure, in such case I guess you could also write this as tests directly in the demo application I guess, but this would also be rather cumbersome. Just as a point to consider.

Just a side note: you can use one demo application to test multiple add-ons in a monorepo, this can be a strength instead of cumbersome

**Comment by gossi** [comment]
I wrote about this a while back [Thoughts on Improving Developer Ergonomics for emberjs](https://gos.si/blog/thoughts-on-improving-developer-ergonomics-for-emberjs/).

I'm super much in favor of this. I'm more exactly interested in their own packages for the reasons @jenweber listed here. Also monorepo setup might make sense (keep in mind: there are monorepos containing addons - they would in turn become a monorepo of monorepos - monorepoception 😱)

**Comment by jenweber** [comment]
A monorepo structure and apps people can name themselves would be awesome. I have seen several addons that would have benefited from having test apps and docs apps together in the same repo as their addon.

I think there are two unknowns that would be good to explore - how much work would a rename/alias be, and could it be mostly done by community contributors? Versus, how much work is a restructure? (A restructure would involve a lot of work for a lot of people.)

If the rename could be done quickly by a small group of community contributors, in a non-breaking way, and the restructure would take a very long time, maybe the next step is that both these ideas should become their own RFCs.

**Comment by knownasilya** [comment]
>  (keep in mind: there are monorepos containing addons - they would in turn become a monorepo of monorepos - monorepoception 😱)

I was thinking that `ember addon my-addon` would create a new repo with a monorepo setup and place the addon in something like `packages/my-addon` in that repo and also create `packages/example-usage` so that it's not mono repo in mono repo. If you wanted to add another addon you could do `ember addon another` inside the repo packages and it would add it there. Same for an app, but something like `ember new docs`.

**Comment by jelhan** [comment]
I agree that dummy app is not a good term. But wondering if monorepo is the best solution. I like the flexibility of a monorepo. But it also adds to the learning curve.

If we go with monorepo I would prefer to have public addons and demo / testing / documentation projects in different folders. Having everything in the same folder can make it difficult to find the public addon within all testing projects. See Ember Auto Import as an example: https://github.com/ef4/ember-auto-import/tree/master/packages

If you need to test a couple of build related scenarios I prefer creating the projects on the fly as part of the tests. I fear the costs of maintaining multiple testing projects. Also it's difficult to see what each project is about due to amount of boilerplate code. [Ember CLI Addon Tests](https://github.com/tomdale/ember-cli-addon-tests) supports that strategy. I started a modernized reimplementation as [Ember Addon Tests](https://github.com/jelhan/ember-addon-tests) some time ago. But still early alpha.

**Comment by lougreenwood** [comment]
IMHO, the justifications to choose more descriptive name simply because the term "dummy" is a non-descriptive colloquialism is enough of a reason. Do we need to "politicise" it (albeit with good intentions)? I worry it distracts the discussion...

**Comment by jenweber** [comment]
Considering the human emotional impact is something that can help us understand where this falls among the many things that could be priorities. There are 100 different ways we could work on for inclusive tech, and there are only so many hours in a day. Strategically, is this a good thing to do now? Later? What helps us decide? I am glad inclusive language is part of the discussion.

If this was solely about having a more programmatically descriptive name, it is probably not worth the effort when we could all be doing other more impactful things.

**Comment by sandstrom** [comment]
I don't think this crusade against words is wise. Banning dummy is just censorship and just generally a bad idea for many reasons:

## Banning dummy won't increase the chances of low IQ people
Someone with a low IQ won't magically get a better chance of having a career in tech just because we don't use the word dummy in our code. The common argument against "ableist language" is that people are excluded from a sector because of the words we use. But the problem for people with low IQ getting a career in highly technical fields is simply that they aren't able to do the work.

## Where does it stop?
That there exist words to describe various maladies is useful, e.g. for treating them. But where does it stop? We ban *dummy* and then *retarded*. How about intellectual disability? Maybe we should ban disability, disabled and also enabled since that's offensive to anyone who isn't fully enabled? 


## Not using words won't make the problem go away
For example, take the word allergic:

```
adjective
  caused by or relating to an allergy: an allergic reaction to penicillin.
  • [predicative] having an allergy to a substance: one and a half per cent of the population is allergic to bee venom.
  • (allergic to) informal having a strong dislike for: I'm allergic to the hype.
```

I have a bunch of allergies, but it's not like that (minor) disability is alleviated by people not saying "I'm allergic to the hype". Words can have multiple meanings, most words in the dictionary have. Banning every word for allergy won't make allergies go away.

**Comment by NullVoxPopuli** [comment]
I think this discussion could be much more productive if we focus on technical merit and learnability of our word choices.. otherwise we're just going to be going in circles. 😢 

Personally, I want this out of the test apps:
 - I want multiple of them
 - They should be named something that makes sense, has little to no ambiguity in meaning 

My associations with the following words (based on colloquial usage):
 - **dummy** - dumb person, semi-flexible mannequin thing used in car crash tests
 - **test app** - used for testing, an app, not an addon - app and addon are already very clearly defined in ember
 
Tangentially, I also propose this folder convention, which means we don't even have a name to talk about! :D 
```
addon/
# ...
tests/
  apps/  # <-- full apps with their own test directories
     {any name here}
     {any name here}
```
and `ember test` / `/tests` would run addon tests and with tests from each app as well
tests could maybe run in an iframe and report test results back to the main tests window via standard iframe communication (postmessage, etc)

**Comment by hjdivad** [comment]
@NullVoxPopuli At the point that we have multiple test apps under `test/apps` we're almost at monorepos.  Do you think there's marginal value in a `test/apps` approach as compared to a monorepo?  For addons that only want multiple test applications the additional overhead is minor but for addons that want to test or demo something about the build, having the test applications have their own `package.json` is valuable.

**Comment by gossi** [comment]
No, we aren't at monorepos then. Imagine this: There is a _monorepo_ with multiple addons, each of them with multiple apps in them to test them. Now think, these would flatten into the _parent_ monorepo. So keep _project_ concerns within them.

I'm wondering what a good structure of multiple _projects_ (apps) within an ember _project_ will look like. By project I mean app, addon and engine - and each of them can have multiple apps (or even projects? maybe only apps) have within them (demo, docs, tests) - I'd rather think into that direction, rather then only looking for tests (unless it shows, they _must_ be separated).

**Comment by hjdivad** [comment]
@gossi I think we're using the term `monorepo` a little differently.

Let me rephrase: at the point that we have multiple test apps under `apps/` we can add `package.json`s so them and use yarn workspaces to let these test apps *also* test and demo build setups that are currently a little annoying to test and illustrate in an addon today (e.g. if you want to test highlander semantics for your addon).

**Comment by wagenet** [comment]
I'm closing this due to inactivity. This doesn't mean that the idea presented here is invalid, but that, unfortunately, nobody has taken the effort to spearhead it and bring it to completion. Please feel free to advocate for it if you believe that this is still worth pursuing. Thanks!

**Comment by NullVoxPopuli** [comment]
This issue will also kinda phase itself out with v2 addons. (eventually)
